openpgp/v2/keys_test.go

```diff
@@ -2022,3 +2021,75 @@ NciH07RTRuMS/aRhRg4OB8PQROmTnZ+iZS0=
 		t.Fatal(err)
 	}
 }
+
+func TestAllowAllKeyFlagsWhenMissing(t *testing.T) {
+	// Make a master key.
+	entity, err := NewEntity("Golang Gopher", "Test Key", "no-reply@golang.com", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	config := &packet.Config{}
+
+	primarySelfSignature, err := entity.VerifyPrimaryKey(time.Now(), config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !entity.PrimaryKey.PubKeyAlgo.CanEncrypt() ||
+		!entity.PrimaryKey.PubKeyAlgo.CanSign() {
+		t.Fatal("PubKeyAlgo must be valid for signature and encryption")
+	}
+
+	/// Flags valid, but not set.
+	primarySelfSignature.FlagsValid = true
+	primarySelfSignature.FlagSign = false
+	primarySelfSignature.FlagCertify = false
+	primarySelfSignature.FlagEncryptCommunications = false
+
+	if isValidSigningKey(primarySelfSignature, entity.PrimaryKey.PubKeyAlgo, config) {
+		t.Error("isValidSigningKey must be false")
+	}
+
+	if isValidEncryptionKey(primarySelfSignature, entity.PrimaryKey.PubKeyAlgo, config) {
+		t.Error("isValidSigningKey must be false")
+	}
+
+	if isValidCertificationKey(primarySelfSignature, entity.PrimaryKey.PubKeyAlgo, config) {
+		t.Error("isValidSigningKey must be false")
+	}
+
+	/// Flags not valid, but set.
+	primarySelfSignature.FlagsValid = false
+	primarySelfSignature.FlagSign = true
+	primarySelfSignature.FlagCertify = true
+	primarySelfSignature.FlagEncryptCommunications = true
+
+	if isValidSigningKey(primarySelfSignature, entity.PrimaryKey.PubKeyAlgo, config) {
+		t.Error("isValidSigningKey must be false")
+	}
+
+	if isValidEncryptionKey(primarySelfSignature, entity.PrimaryKey.PubKeyAlgo, config) {
+		t.Error("isValidSigningKey must be false")
+	}
+
+	if isValidCertificationKey(primarySelfSignature, entity.PrimaryKey.PubKeyAlgo, config) {
+		t.Error("isValidSigningKey must be false")
+	}
+
```

Comment:
This scenario can't actually happen in a parsed key, because the `FlagsValid` really just means whether the flags were present in that signature packet. It's not the end of the world to test it anyway of course, but might be slightly confusing, so I'd tend to just remove this one.

```suggestion
```